### PR TITLE
Code cleanup and FederatedZkClient Multi test.

### PR DESCRIPTION
**Pull request by Marcosrico**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

https://github.com/apache/helix/issues/2240

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

To test implementation of FederatedZkClient Multi, tests must be written. Mainly rearranged current test base for cleaner code and added specific test for FederatedZkClient multi support. This pull request is part 1 of 2 for implementing transaction support in zk client.

### Tests

- [x] The following tests are written for this issue:

Solely rearranged tests under /multizk and combined similar code to clean up the tests. Wrote separate test to test future implementation of FederatedZkClient ZkMulti in testMultiInMultiZk. Currently FederatedZkClient ZkMulti is not supported.

- [x] The following is the result of the "mvn test" command on the appropriate module:

```
[INFO] Tests run: 1320, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11,288.389 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1320, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.6:report (generate-code-coverage-report) @ helix-core ---
[INFO] Loading execution data file /Users/mapeng/helix/helix-core/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Core' with 928 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:08 h
[INFO] Finished at: 2022-12-21T22:09:49+01:00
[INFO] ------------------------------------------------------------------------
```

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import org.apache.helix.zookeeper.api.client.RealmAwareZkClient;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
+import org.apache.helix.zookeeper.impl.factory.DedicatedZkClientFactory;
+import org.apache.helix.zookeeper.zkclient.ZkServer;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+
+import java.util.*;
+
```
Comment (xyuanlu) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java:43]: Please add desiccation here, # of zk servers we are starting etc.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkHelixJavaApis.java [hunk 2]
```diff
@@ -440,13 +323,8 @@ public void testZKHelixManagerCloudConfig() throws Exception {
...
     _zkHelixAdmin.addInstance(clusterName, instanceConfig);
 
-    RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder connectionConfigBuilder =
-        new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder();
-    // Try with a connection config without ZK realm sharding key set (should fail)
-    RealmAwareZkClient.RealmAwareZkConnectionConfig invalidZkConnectionConfig =
-        connectionConfigBuilder.build();
-    RealmAwareZkClient.RealmAwareZkConnectionConfig validZkConnectionConfig =
-        connectionConfigBuilder.setZkRealmShardingKey("/" + clusterName).build();
+    createZkConnectionConfigs(clusterName);
```
Comment (xyuanlu) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkHelixJavaApis.java:326]: We dont need to call this again. It is already populated in previous test

File: helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                    if (stateMachine != null) {
+                        StateModelFactory stateModelFactory = stateMachine.getStateModelFactory("Task");
+                        if (stateModelFactory instanceof TaskStateModelFactory) {
+                            ((TaskStateModelFactory) stateModelFactory).shutdown();
+                        }
+                    }
+                });
+            }
+
+            if (!MOCK_PARTICIPANTS.isEmpty()) {
```
Comment (xyuanlu) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java:125]: what's the difference here..?

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkConnectionConfig.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        // Register task state model for task framework testing in later methods
+        Map<String, TaskFactory> taskFactoryReg = new HashMap<>();
+        taskFactoryReg.put(MockTask.TASK_COMMAND, MockTask::new);
+        // Register a Task state model factory.
+        StateMachineEngine stateMachine = mockNode.getStateMachineEngine();
+        stateMachine
+                .registerStateModelFactory("Task", new TaskStateModelFactory(mockNode, taskFactoryReg));
+
+        mockNode.syncStart();
+        MOCK_PARTICIPANTS.add(mockNode);
```
Comment (xyuanlu) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkConnectionConfig.java:149]: If we have `MOCK_PARTICIPANTS` as a non-static var, we probably don't need to repeat this test in 2 classes?

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkConnectionConfig.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  /**
+   * Test Helix Participant creation and addition.
+   * Helix Java APIs tested in this method are:
+   * ZkHelixAdmin and ZKHelixManager (mock participant/controller)
+   */
+  @Test(dependsOnMethods = "testCreateClusters")
+  public void testCreateParticipants() throws Exception {
+    // Create two ClusterSetups using two different constructors
+    // Note: ZK Address here could be anything because multiZk mode is on (it will be ignored)
+    //HelixAdmin helixAdminZkAddr = new ZKHelixAdmin(ZK_SERVER_MAP.keySet().iterator().next());
```
Comment (mgao0) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkConnectionConfig.java:100]: I'm curious if there a reason to have this comment? Also saw a few cases like this in the rest of the lines.

**Review by mgao0 — COMMENTED** [review]
Thanks for creating MultiZkTestBase, this is an example of great craftsmanship. Your tests are straightforward and LGTM. Will try to understand more about the multi-zk setup and leave more comments if any.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkConnectionConfig.java (hunk 5 above)
Comment (Marcosrico) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkConnectionConfig.java:100]: There is no reason for these comments, thank you for pointing it out! Both test cases in TestMultiZkConnectionConfig and TestMultiZkHelixJavaApis were very similar (the reason for this PR). I'm guessing they were copy pasted and commented out for a specific purpose and went under my radar when creating the the testbase.

**Comment by qqu0127** [comment]
nit: I think the coding style is indent with 2 white spaces, not tab. Let's make them consistent. Thanks!

File: helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    protected MockMetadataStoreDirectoryServer _msds;
+    protected static final Map<String, Collection<String>> _rawRoutingData = new HashMap<>();
+    protected RealmAwareZkClient _zkClient;
+    protected HelixAdmin _zkHelixAdmin;
+
+    // Save System property configs from before this test and pass onto after the test
+    protected final Map<String, String> _configStore = new HashMap<>();
+
+    protected static final String ZK_PREFIX = "localhost:";
+    protected static final int ZK_START_PORT = 8977;
```
Comment (rahulrane50) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java:65]: nit: Just curious for the port number used. I have mostly seen port number being used in range of 21xx or 129xx range.

**Review by rahulrane50 — COMMENTED** [review]
Overall very nice refactoring! Great work!

File: helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java (hunk 6 above)
Comment (Marcosrico) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java:65]: Just to differentiate them with single Zk server setup which uses the ports you mentioned above

File: helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+import org.apache.helix.zookeeper.api.client.HelixZkClient;
+import org.apache.helix.zookeeper.api.client.RealmAwareZkClient;
+import org.apache.helix.zookeeper.datamodel.serializer.ZNRecordSerializer;
+import org.apache.helix.zookeeper.impl.factory.DedicatedZkClientFactory;
+import org.apache.helix.zookeeper.zkclient.ZkServer;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+
+import java.util.*;
```
Comment (NealSun96) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java:42]: In general we do specific imports instead of `java.util.*`.

**Review by NealSun96 — CHANGES_REQUESTED** [review]

File: helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java (hunk 7 above)
Comment (Marcosrico) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/MultiZkTestBase.java]: Addressed, thanks!

File: zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestFederatedZkClient.java [hunk 8]
```diff
@@ -20,13 +20,7 @@
...
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.NoSuchElementException;
+import java.util.*;
```
Comment (qqu0127) [inline zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestFederatedZkClient.java:23]: let's try not to use wildcard import, thanks.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiInMultiZk.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                            .setRoutingDataSourceEndpoint(_msdsEndpoint + "," + ZK_PREFIX + ZK_START_PORT)
+                            .setRoutingDataSourceType(RoutingDataReaderType.HTTP_ZK_FALLBACK.name()).build(),
+                            new RealmAwareZkClient.RealmAwareZkClientConfig());
+            _zkClient.setZkSerializer(new ZNRecordSerializer());
+        } catch (Exception ex) {
+            for (StackTraceElement elm : ex.getStackTrace()) {
+                System.out.println(elm);
+            }
+        }
+        System.out.println("end start");
```
Comment (qqu0127) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiInMultiZk.java:61]: I don't think this is necessary and meaningful, let's remove it.

**Review by qqu0127 — COMMENTED** [review]
Overall looks good, left some minor comments.

**Comment by junkaixue** [comment]
Please make sure the PR follow the PR template including descriptions, linked issues and paste test results: https://github.com/apache/helix/wiki/Pull-Request-Description-Template

**Review by qqu0127 — APPROVED** [review]
LGTM, thanks for the cleanup!

**Review by xyuanlu — APPROVED** [review]
LGTM. Thanks for making the change.

**Comment by Marcosrico** [comment]
This PR is ready to merge, approved by @qqu0127 @xyuanlu 
Commit message:
Code cleanup and refactoring for FederatezZkClient Multi implementation.

**Review by NealSun96 — APPROVED** [review]
LGTM

**Pull request merged**